### ISSUE TITLE
Code for Figure1 (a)

### ISSUE THREAD

**Issue by Irish-kw** [needylove/OrdinalEntropy#2]
HI:

It is possible to obtain code for Figure1 (a) ?

Thank you

**Comment by needylove** [comment]
Hi Irish-kw,

We use below the function, entropy_summary,  to estimate the entropy of feature spaces for Figure 1 (a)

Thanks.


def entropy_summary(features):
    with torch.no_grad():
        n, c, h, w = features.size()
        features = F.interpolate(features, size=[h // 8, w // 8], mode='nearest')
        features = features.permute(0, 2, 3, 1)  # n, h, w, c
        features = torch.flatten(features, start_dim=1, end_dim=2)

        _entropy_original = 0
        _entropy_normalized_original = 0
        _entropy_normalized_l2 = 0
        _entropy_normalized_l2_square = 0
        _entropy_normalized_l2_weighted = 0
        for i in range(n):
            p = F.normalize(features[i, :], dim=1)
            _dist = F.pdist(p)
            p_n, p_d = p.size(0), p.size(1)

            # _entropy = (p_d / (p_n * (p_n - 1))) * torch.sum(_dist)
            # _entropy_normalized_original = _entropy_normalized_original + _entropy
            _dist_normalized = _dist.pow(2)
            _dist_normalized = _dist.log()
            _entropy_normalized_original = _entropy_normalized_original + (p_d / (p_n * (p_n - 1))) * torch.sum(_dist_normalized)

            _entropy_normalized_l2 = _entropy_normalized_l2 + torch.mean(_dist)
            _entropy_normalized_l2_square = _entropy_normalized_l2_square + torch.mean(_dist.pow(2))
            _max = torch.max(_dist)
            _min = torch.min(_dist)
            _wt = ((_dist - _min) / _max)
            _entropy_normalized_l2_weighted = _entropy_normalized_l2_weighted + torch.mean(_dist * _wt)

            _dist = F.pdist(features[i, :])
            _dist = _dist.pow(2)
            _dist = _dist.log()
            _entropy_original = _entropy_original + (p_d / (p_n * (p_n - 1))) * torch.sum(_dist)

        # e11 = _entropy_original.detach()
        # e11 = _entropy_original.cpu()
        # e11 = _entropy_original.detach().numpy()
        e1 = _entropy_original.detach().cpu().numpy() / n
        e2 = _entropy_normalized_original.detach().cpu().numpy() / n
        e3 = _entropy_normalized_l2.detach().cpu().numpy() / n
        e4 = _entropy_normalized_l2_square.detach().cpu().numpy() / n
        e5 = _entropy_normalized_l2_weighted.detach().cpu().numpy() / n

    return e1,e2,e3,e4,e5

**Comment by Irish-kw** [comment]
Thank you!

**Comment by Irish-kw** [comment]
I'm sorry that I want to ask again, e1 is Equation1, and e3 is Equation1 with normalized by norm2, so I think e5 is Equation 8, right? thank you for your help.

**Comment by needylove** [comment]
Correct. Apart from Figure 1(a), you can find more results in the Appendix, Figure A, where estimated entropy through e1, e3, and e5 are reported.

**Comment by Irish-kw** [comment]
Thank you for your reply ! 
But I'm not sure why the result(value) can be negative, Appendix B Figure A (d), (e), (f).
Based on my knowledge, the entropy in information theory should be between 0 and 1.
What am I missing?

Thank you for your help!

**Comment by needylove** [comment]
We estimate the entropy with the MeanNN estimator, which approximates P(Z) with the distance between samples. However, the scale of distance varies a lot, while the true probability is always between 0 to 1. Assuming given the distances scale between (0, 1), then, the estimated entropy will be negative due to the log operation in the MeanNN estimator.  This scale ambiguity is one of the reasons why we report results w/o normalizing feature vectors.

**Comment by Irish-kw** [comment]
「Assuming given the distances scale between (0, 1)」
Great ! this is what I needed to know!

Thank you !!!!!

**Comment by Irish-kw** [comment]
HI: 

Sorry for asking the question again:

In the paper Appendix B:

「We modify the last layer of a ResNet-50 model to a convolution operation
with kernel size 1 × 1, and train the modified model with MSE as regression model 」

I comprehensive that before the convolution layer is z after the convolution layer is y_hat
Hence, I use z which is before the convolution layer as features.

In Equation 8 is w_ij = ||y_i - y_j||_2
but in the following code, It seems w_ij is _wt.
and why calculate _wt at z feature space (_dist ), _wt = ((_dist - _min) / _max) ?
Am I missing something? 
Thank you for your help!!

p = F.normalize(features[i, :], dim=1)
_dist = F.pdist(p)
_max = torch.max(_dist)
_min = torch.min(_dist)
_wt = ((_dist - _min) / _max)
_entropy_normalized_l2_weighted = _entropy_normalized_l2_weighted + torch.mean(_dist * _wt)
